src/modm/math/geometry/vector2.hpp

```diff
@@ -12,373 +13,409 @@
+
+/**
+ * \brief	Class for handling common vector operations (2D)
+ *
+ * Operations:
+ * \code
+ *	+ : addition of points
+	*	- : different of points
+	*	* : dot product or scalar multiplication
+	*	/ : scalar division
+	*	^ : cross product (determinant)
+	*	~ : perpendicular
+	* \endcode
+	*
+	* Adapted from the implementation of Gaspard Petit (gaspardpetit@gmail.com)
+	* and heavily modified.
+	*
+	* \see <a href"http://www-etud.iro.umontreal.ca/~petitg/cpp/point.html">Homepage</a>
+	*
+	* \author	Fabian Greif
+	* \author	Niklas Hauser
+	* \author	Thomas Sommer
+	* \ingroup	modm_math_geometry
+	*/
+template<typename T>
+class Vector<T, 2>
+{
+	using VecT1 = Vector<T, 1>;
+	using VecT2 = Vector;
+
+public:
+	using WideType = GeometricTraits<T>::WideType;
+	using FloatType = GeometricTraits<T>::FloatType;
+
+	T x{0}, y{0};
+
+	// fundamental constructors
+	constexpr Vector() = default;
+
+	constexpr explicit Vector(T xy)
+		: x(xy), y(xy) {}
+
+	template<typename U>
+	requires std::integral<T> && std::floating_point<U>
+	constexpr explicit Vector(T xy)
+		: x(std::round(xy)), y(std::round(xy)) {}
+
+	constexpr Vector(T x, T y)
+		: x(x), y(y) {}
+
+	template<typename U>
+	requires std::integral<T> && std::floating_point<U>
+	constexpr Vector(U x, U y)
+		: x(std::round(x)), y(std::round(y)) {}
+
+	template<typename U>
+	constexpr Vector(Vector<U, 2> v) : Vector(v.x, v.y) {}
+
+	// vector constructors
+	constexpr Vector(VecT1 vx, VecT1 vy) : Vector(vx.x, vy.x) {}
+	constexpr Vector(T x, VecT1 vy) : Vector(x, vy.x) {}
+	constexpr Vector(VecT1 vx, T y) : Vector(vx.x, y) {}
+
+	// matrix constructors
+	constexpr Vector(const Matrix<T, 2, 1> &rhs)
+		: Vector(reinterpret_cast<const T*>(&rhs)[0], reinterpret_cast<const T*>(&rhs)[1]) {}
+
+	// array constructors
+	template<typename U>
+	constexpr explicit Vector(U *array) : Vector(array[0], array[1]) {}
+
+	// as matrix convertors
+	Matrix<T, 2, 1>&
+	asMatrix() { return *((Matrix<T, 2, 1> *) this); }
+
+	const Matrix<T, 2, 1>&
+	asMatrix() const { return *((Matrix<T, 2, 1> *) this); }
+
+	Matrix<T, 1, 2>&
+	asTransposedMatrix() { return *((Matrix<T, 1, 2> *) this); }
+
+	const Matrix<T, 1, 2>&
+	asTransposedMatrix() const { return *((Matrix<T, 1, 2> *) this); }
+
+	// getters
+	T getX() const { return x; }
+	T getY() const { return y; }
+
+	// setters
+	void setX(T x) { this->x = x; }
+
+	template<typename U>
+	requires std::integral<T> && std::floating_point<U>
+	void setX(U x) { this->x = std::round(x); }
+
+
+	void setY(T y) { this->y = y; }
+
+	template<typename U>
+	requires std::integral<T> && std::floating_point<U>
+	void setY(U y) { this->x = std::round(y); }
+
+
+	void set(T x, T y) {
+		this->x = x;
+		this->y = y;
+	}
+
+	template<typename U>
+	requires std::integral<T> && std::floating_point<U>
+	void set(U x, U y) {
+		this->x = std::round(x);
+		this->y = std::round(y);
+	}
+
+	// accessors
+	T& operator[] (std::size_t index)
+	{ return reinterpret_cast<T*>(this)[index]; }
```

Comment:
That code is undefined behaviour. You can actually cast a pointer of a standard-layout type to the pointer type of its first member and dereference it but the array indexing into members is not allowed. I tried many things to exactly keep the classinterface the same and make it defined but all these solutions had bad code gen.

A possible way to go could be to store the data as an array and provide member function accessors for x and y that return a reference. Users would need to change `.x` to `.x()` though. We could make basically everything `constexpr` in here. I would not feel comfortable to ship the UB version in producion code.

 @salkinium Do you think this change in interface would be acceptable?